## EDA- Elobrotary Data Analysis and Data Wrangling techniques

### 1- Import Data

In [1]:
# Libraries:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
## Importing the dataset
df= sns.load_dataset('titanic')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 2- Checking and exploring data. 

In [3]:
# columns id are known as variables
# columns values are known as observations
# columns dtypes are known as data types
# object = string
#catogorical = int, float
df.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [4]:
df.info()
# check null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [5]:
# check rangeindex = 891 entries. 0 to 890
# age, cabin, embarked have null values
# age has 714 non-null float64 =891-714 = 177 null values
# cabin has 204 non-null object= 891-204 = 687 null values
# embarked has 889 non-null object= 891-889 = 2 null values
df.isnull().sum()


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [6]:
df1=df

In [7]:
df1

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [8]:
df1.isnull().sum()/len(df1)*100
# checking percentage of null values

survived        0.000000
pclass          0.000000
sex             0.000000
age            19.865320
sibsp           0.000000
parch           0.000000
fare            0.000000
embarked        0.224467
class           0.000000
who             0.000000
adult_male      0.000000
deck           77.216611
embark_town     0.224467
alive           0.000000
alone           0.000000
dtype: float64

## How to deal with missing values in EDA analysis?
## cataogorical /object, numeric and boolean
### Handling missing values is a crucial step in Exploratory Data Analysis (EDA). Let's break it down by data type:
## Remember : Catagorical values means yes/no, male/female in this case we cant take mean we can take mode.

1. **Numeric Values:**
   - **Imputation:** Replace missing numeric values with the mean, median, or mode of the column. This helps to maintain the overall distribution.
   - **Flagging:** Introduce a new boolean column indicating whether the value was missing in the original column. This can provide additional information.

2. **Categorical/Object Values:**
   - **Mode Imputation:** Replace missing categorical values with the mode (most frequent category).
   - **Creating a New Category:** Introduce a new category (e.g., "Unknown" or "Other") to represent missing values.
   - **Use Predictive Models:** Train a model to predict missing categorical values based on other features.

3. **Boolean Values:**
   - **Imputation:** Since boolean values are typically 0 or 1, you might replace missing values with the most frequent value (0 or 1).
   - **Flagging:** Similar to numeric values, you can create a new boolean column indicating whether the original value was missing.

Always remember to document and justify your choice of imputation method, as it can impact the results of your analysis.

Would you like more information on any specific aspect of handling missing values?

In [9]:
# values in age column
df1['age'].unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

In [10]:
df1['age'].max()

80.0

In [11]:
df1['age'].min()

0.42

## 1st Method- Replacing Nan values with mean.


In [12]:
# replace null values with mean in age column
df1['age'].isnull().sum()

177

In [13]:
print (df1['age'].max())
print (df1['age'].min())
print (df1['age'].mean())

80.0
0.42
29.69911764705882


In [14]:
df1['age'].fillna(df1['age'].mean(), inplace=True)
df1

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone  
0     Third    man        True  NaN  Southampton    no  False  
1     First  woman       False    C    Cherbourg   yes  False  
2     Third  woman       False  NaN  Southampton   yes   True  
3     First  woman       False    C  Southampton   yes  False  
4     Third    man        True  NaN  Southampton    no   True  
..      ...    ...         ...  ...          ...   ...    ...  
886  Second    man        True  NaN  Southampton    no   True  
887   First  woman       False    B  Southampton   yes   True  
888   Third  woman       False  NaN  Southampton    no  False  
889   First    man        True    C    Cherbourg   yes   True  
890   Third    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [15]:

# df1 has no null values replaced with mean
df1.isnull().sum()


survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [16]:
df2=df1

In [17]:
df2.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## 2nd method- If there is many nan values so we will drop the coulmn.
## e.g; deck

In [18]:
## now we remove nan from deck column.
## as there is many nan values so we drop the coulmn.
df2['deck'].unique()

[NaN, 'C', 'E', 'G', 'D', 'A', 'B', 'F']
Categories (7, object): ['A', 'B', 'C', 'D', 'E', 'F', 'G']

In [19]:
## dropping deck column
df2=df2.drop(columns=["deck"])

In [20]:
## after dropping we are checking null values.
df2.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

## Method 3- Replacing values with mode

In [21]:
## now we replace nan values with mode in embarked and embarked town column.
df3=df2

In [22]:
df3['embarked'].fillna(df3['embarked'].mode()[0], inplace=True)
df3['embark_town'].fillna(df3['embark_town'].mode()[0], inplace=True)
df3

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male  embark_town alive  alone  
0     Third    man        True  Southampton    no  False  
1     First  woman       False    Cherbourg   yes  False  
2     Third  woman       False  Southampton   yes   True  
3     First  woman       False  Southampton   yes  False  
4     Third    man        True  Southampton    no   True  
..      ...    ...         ...          ...   ...    ...  
886  Second    man        True  Southampton    no   True  
887   First  woman       False  Southampton   yes   True  
888   Third  woman       False  Southampton    no  False  
889   First    man        True    Cherbourg   yes   True  
890   Third    man        True   Queenstown    no   True  

[891 rows x 14 columns]

In [23]:
# there is no null values in df3
df3.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

### Summary:( Data Wrangling)(EDA)
1.  Importing Libraries.
2.  Import Dataset.
3.  Explore Data.
    *   Infromation
    *   Datatypes
    *   Missing Values
    *   Take sense of your data.
4.  Understanding variables.
5.  Relationship between the variables(Heatmap, paiplot, coorelation)
6.  Brainstroming
    * Normalize ( Decomposition of Big data to small data like coorelation in 1(0-1))
    * Removing outliers ( if one values is affecting the whole data we will remove e.g; avergae heigh=5.7, alam channa=14.5) 
7. Tidy Data or Clean Data ready to use.
8. Ready for statistical Analysis.
9. Ready for prediction.
10. Ready for machine learning.
11. Ready for Deep Learning.

#### Data Science Rule:
80% of time consumed in step 1-6.
20% of time used for remaing steps.
